Activity Changelog UI

jeffpamer opened the pull request:
This adds a first-pass UI for the activity directive changelog. The 10 most recent modifications to an activity directive are stored in a rolling list in the database. Clicking the changelog in the activity directive form header allows viewing a list of changelog entries with diffs displayed (when possible) between each revision. From there the user can either directly restore a revision, or preview the differences between that revision and the current one and decide whether to restore from there.

![2023-09-28 14 53 21](https://github.com/NASA-AMMOS/aerie-ui/assets/888818/b271eb09-0ea7-4ef9-8c60-4f99f5f66f6f)

There are some edges cases and strange scenarios we may likely need to iterate on from here. For instance:
- I found some scenarios where I couldn't determine any changes between two revisions. Such as when creating a new preset from the activity's current parameters. This causes an entry to be written to the changelog, but no parameters change.
- At the moment it doesn't seem like we have a way of differentiating multiple changes occurring as a result of using the API, or selecting a preset, etc. We may need to look into storing preset selection as one of the changes if we want to call that out specifically in the changelog.

jeffpamer committed:
Rough first pass at fetching changelog and displaying results per activity

jeffpamer committed:
Basic working UI for change log, no restore or preview

jeffpamer committed:
Account for anchor differences

jeffpamer committed:
small formatting fixes and type corrections

jeffpamer committed:
Catch scenarios with no changes, e.g. creating a preset

jeffpamer committed:
remove unused data, format durations

jeffpamer committed:
move changelog button into activity name header

jeffpamer committed:
wire up effect for restoring from changelog

jeffpamer committed:
Close the changelog if the selected activity changes

jeffpamer committed:
use new changelog icon

jeffpamer committed:
Basic working revision preview

jeffpamer committed:
Add header and 'restore' action to revision preview

jeffpamer committed:
more efficient effective arguments construction

jeffpamer committed:
add Last Modified By to activity directive form

jeffpamer committed:
Add permission handling

jeffpamer committed:
fix linter errors

jeffpamer committed:
drive anchor offset detection from props instead of activity object

jeffpamer committed:
Fix tests. Add basic handling of array/struct types

AaronPlave commented:
> I found some scenarios where I couldn't determine any changes between two revisions. Such as when creating a new preset from the activity's current parameters. This causes an entry to be written to the changelog, but no parameters change.

Is this change revertible? And if not should it even be stored in the changelog?

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
```diff
@@ -0,0 +1,365 @@
+        <div class="change-summary st-typography-body">
+          {activityRevisionChangeMap[i].name}
+        </div>
+        <div class="changed-by st-typography-label">{revision.changed_by}</div>
```
Might want to add an `@` prefix to user to match simulation history and plan snapshot @lklyne 
<img width="418" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/f15517c7-dc0b-4ce9-a8ae-85e7d9bb6dc4">

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
```diff
@@ -0,0 +1,365 @@
+    // Manually check remaining fields that could have changed and require extra formatting
+
+    if (current.start_offset !== previous.start_offset) {
+      differences['Start Time'] = {
```
One thing to ask users about might be time formatting here - do they care about the specific time values or are they more interested that the field changed? Have heard comments before that the time offset is not particularly understandable vs a more human readable format.
<img width="116" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/83d6ad32-aae0-49df-b2c2-c19b50b9a279">

jeffpamer replied:
Taking another look, it looks like the designs specify displaying these in DOY format so I can update that. It is a good question though of which is ultimately more useful information.

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
Managed to get into a state where I have an entry with "No Changes" that, when previewed, lists 2 changes. Was testing this with ParameterTest in Bananation and was incrementing primitiveShort and changing primitiveBoolean. Was using arrow keys to spam number changes which may or may not be relevant, will continue investigating. 
<img width="573" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/7948d5a7-7f3c-4c8b-ab98-ad75e22be544">
<img width="286" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/81819887-0be8-4147-bac6-6d983500179b">

AaronPlave replied:
Ah I think I may not have understood the changelog behavior. Perhaps this is explained by _some_ edge case causing "No Changes" to appear and the preview is the diff between the current state and this point in time?

AaronPlave replied:
Can still get "No Changes" to appear when spamming the integer input, plus weird updates happen to the number input where the UI might be getting the old set of values which lag behind the dirty input..

jeffpamer replied:
Yeah I was surprised to run into this as well, but it definitely seems like there are potential edge cases where we get records written to the changelog with no discernible difference between two revisions. I thought about filtering those out of the list entirely, but it seemed kind of odd to do that when were already only showing 10 revisions. I'm not sure what else to do here unless the backend has some logic to be more strict about only writing changelog entries when something changes?

AaronPlave replied:
@skovati would you happen to know what could be behind this type of changelog entry?

skovati replied:
Hmm, I was able to reproduce the "No Changes" entry when spamming the integer input. The backend for this is intentionally indifferent about the actual changes that occur, it simply copies the state of a given `activity_directive` row on any `update` statement. Is it possible that when spamming input fast enough, the UI sends two GraphQL mutations to "update" to the same number?

AaronPlave replied:
Interesting, yeah that could be possible. I think we should also be debouncing those inputs anyway which may for the most part resolve this issue.

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
```diff
@@ -0,0 +1,365 @@
+      : { currentValue: '', previousValue: '' };
+
+    if (changedProperties.length > 1) {
+      return { currentValue: `${changedProperties.length} Changes`, name: 'Multiple Changes', previousValue: '' };
```
Curious if users will want to be able to see each change but suppose they can just preview this?
<img width="553" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/e4f2e3e5-c5ec-4276-b5a4-895e86acb66b">

jeffpamer replied:
Yeah I was kind of wondering if we could show more changes in a tooltip or some other way, but I think we'll always hit _some_ effective limit of what we'll be able to display. Theoretically through the API or something we might have a changelog entry with 10+ changes if not more?

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
Seeing a recurrence of the input blur submission issue that I cannot seem to find in the UI GH issues. Not something introduced by this PR but something that may come up when using this functionality from time to time? 

https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/ae1b9928-da13-4418-adc4-16c4afb3ca7e

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
Missing bottom padding
<img width="508" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/cf01fda3-596e-43be-8fda-15dea1a5631c">

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
```diff
@@ -0,0 +1,365 @@
+
+      // At some point there will be no previous revision to compute a difference from
+      if (!previousRevision) {
+        return { currentValue: '', name: 'Earliest Known Revision', previousValue: '' };
```
Does `Earliest Known Revision` mean "oldest stored entry in changelog stack" or "original state" of some sort? To me this feels potentially misleading as the user isn't informed that the changelog is restricted to 10 entries?

AaronPlave replied:
@lklyne

jeffpamer replied:
Earliest Known Revision is the oldest entry we have available in the changelog.

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
Wondering if we need to re-run activity parameter validation on parameter preview? Issue is that when you preview a changelog entry the validation is stale so you end up in situations like this one where your previewed params do not match the validation:
<img width="257" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/13cac922-c3a1-4940-b208-0ce6b1eb18cb">

jeffpamer replied:
This is a great point, definitely need to think about how the validations interact with previewed values.

jeffpamer replied:
Just pushed up a change to re-run the validations with the revision args if we're previewing.

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
Wondering if we need to include a bit of guidance somewhere on how this changelog functions? It's not immediately clear to me what is happening when I hit preview/restore. I think we're showing the directive params at that particular point in time (where the diff on the right of the card represents the latest change at that time)? I think for me there's a bit of a disconnect between the diff on the right and the number of changes that you'd actually be applying if you restore from that point. Just to double check - you're not restoring that value right? You're restoring all changes made up until that point? Wondering if there could be language in the preview mode and/or some sort of confirmation modal for the restoring process that would help make this more concrete? @lklyne

jeffpamer replied:
Yes you're right in that when you click preview (or restore for that matter) you are rewinding the entire change history back to that revision, whereas in the changelog rows themselves we're trying to show a rolling list of what is diff only between one revision and the previous. And I definitely see how that's potentially confusing. I think that was essentially the reason for providing the preview, as it might be hard to visualize what exactly changed over the course of several revisions that give you an option to preview the full state of what the activity will be reverted to.

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
"Restore" here is different from the plan snapshot "restore" – will this cause any confusion? @lklyne

AaronPlave commented on src/components/activity/ActivityDirectiveChangelog.svelte:
```diff
@@ -0,0 +1,365 @@
+  let effectiveRevisionArguments: (ArgumentsMap | undefined)[];
+  $: effectiveRevisionArguments = [];
+
+  function formatDate(dateString: string) {
```
Should use same formatting as the sim and plan snapshot time labels which we almost are here but there's a difference in usage of the comma 
<img width="311" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/f3a2a750-5734-4669-863d-9cddb0e68320">
<img width="135" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/1d949fce-82e0-4a2a-a6fd-050f8dcc6f73">
@lklyne

jeffpamer commented:
Really appreciate your thorough review here @AaronPlave. I'll try to push up some changes for the more immediately actionable items shortly, but I'm unfortunately quite time restricted on how much time I can spend on Aerie this week. We might need to have a discussion about whether these pain points are blockers or can be iterated on later?

jeffpamer committed:
Fix date formatting, username display, and padding

jeffpamer committed:
Display Start Time in DOY format

jeffpamer committed:
re-run validations with preview arguments if supplied

AaronPlave commented:
> Really appreciate your thorough review here @AaronPlave. I'll try to push up some changes for the more immediately actionable items shortly, but I'm unfortunately quite time restricted on how much time I can spend on Aerie this week. We might need to have a discussion about whether these pain points are blockers or can be iterated on later?

Yeah we should tag up about what tweaks, if any, are needed to get this first version out.

duranb commented:
> > I found some scenarios where I couldn't determine any changes between two revisions. Such as when creating a new preset from the activity's current parameters. This causes an entry to be written to the changelog, but no parameters change.
> 
> Is this change revertible? And if not should it even be stored in the changelog?

I don't think that the creation of the preset should be revertible since it's not necessarily plan specific. You're seeing an entry in the changelog probably because upon creation, the UI also sends a query to assign after successfully creating. I think that maybe the assignment should at least be revertible if that data is tracked.

duranb reviewed (changes requested):
I'm seeing a small styling issue with the Restore/Preview buttons being stacked for just the "Earliest" revision. The other revision buttons are 👌🏻 
![Screenshot 2023-10-03 at 3 05 23 PM](https://github.com/NASA-AMMOS/aerie-ui/assets/5290214/b6822cf6-6869-4cb7-9d4e-306e19211318)

duranb commented:
Another suggestion for a different PR is maybe adding some loading state after clicking "Restore". In my plan that has 180 directives, I'm noticing that the directive's value that I restored doesn't get set until a second later after clicking "Restore". For that second I thought that I didn't click the right button.

jeffpamer committed:
use revision when getting effective arguments, use current name of activity in changelog header

jeffpamer committed:
constrain width to always have enough room for restore/preview buttons

jeffpamer commented:
Just pushed up a styling change to ensure that we always have enough width to display the restore/preview buttons side-by-side, and use text overflow ellipses if needed for the values.

![image](https://github.com/NASA-AMMOS/aerie-ui/assets/888818/655a7fb7-0ca3-4340-ae13-44bd7b107b71)

jeffpamer committed:
add more generic method for deep diffing objects

jeffpamer committed:
add tooltip for overflowed values

jeffpamer commented:
Pushed up a couple more changes to more generically handle diffing of objects (so we're recursively comparing them by value), so that we can display a nicer diff of changes nested in more complex parameter types. And also added tooltips for text labels that may be potentially cutoff by overflow.

![image](https://github.com/NASA-AMMOS/aerie-ui/assets/888818/38f67ce3-9cd4-4766-ab3e-cd5d7fb1fb6a)

AaronPlave reviewed (approved):
Looks great, I'm sure we'll have some feedback on it that will help us figure out the next iteration.

duranb commented on src/utilities/permissions.ts:
```diff
@@ -625,6 +625,9 @@ const queryPermissions = {
     );
   },
+  RESTORE_ACTIVITY_FROM_CHANGELOG: (user: User | null): boolean => {
+    return isUserAdmin(user) || getPermission(['restoreActivityFromChangelog'], user);
```
I think this might need to duplicate the permission logic of `UPDATE_ACTIVITY_DIRECTIVE`.

jeffpamer committed:
hook changelog auto-closing to a change in selected ID

jeffpamer committed:
Update restore permissions check to account for plan ownership

jeffpamer commented:
Updated the auto-close logic to look specifically for a changing selected activity ID, so we don't hit that side effect where and update to the current activity also closes the changelog. I also updated the permissions check to consider plan ownership/collaborator.
